src/aspects/ec2-automated-shutdown/index.ts

```diff
@@ -0,0 +1,178 @@
+    /**
+     * Optional custom metric configuration.
+     * If not provided, defaults to CPU utilization with a 5% threshold.
+     */
+
+    readonly metricConfig?: CustomMetricConfig;
+
+    /**
+     * Optional Lambda configuration settings.
+     */
+    readonly lambdaConfiguration?: LambdaConfiguration;
+
+    /**
+     * Optional KMS Encryption Key to use for encrypting resources.
+     */
+    readonly encryption?: IKey;
+}
+
+/**
+ * Aspect that applies a mechanism to automatically shut down
+ * an EC2 instance when its CPU utilization falls below a specified threshold.
+ * Allows for cost optimization and the reduction of resources not being actively used.
+ */
+export class Ec2AutomatedShutdown implements IAspect {
+    private static policyByStack: Map<string, PolicyStatement> = new Map();
+    private static lambdaByStack: Map<string, SecureFunction> = new Map();
+    private readonly defaultMetricConfig: CustomMetricConfig = {
+        namespace: 'AWS/EC2',
+        metricName: 'CPUUtilization',
+        period: Duration.minutes(1),
+        statistic: 'Average',
+        threshold: 5
+    };
+
+    constructor(private readonly props: Ec2ShutdownProps) {}
+
+    visit(node: IConstruct): void {
+        if (!(node instanceof Instance) && !(node instanceof CfnInstance)) {
+            return;
+        }
+        const stack = Stack.of(node);
+        this.createShutdownMechanism(node, stack);
+    }
+
+    private getLambdaFunction(stack: Stack): SecureFunction {
+        const stackId = stack.stackId;
+        if (!Ec2AutomatedShutdown.lambdaByStack.has(stackId)) {
+            const lambda = new SecureFunction(stack, 'Ec2ShutdownFunction', {
+                code: Code.fromAsset(join(__dirname, 'handler')),
+                runtime: Runtime.NODEJS_20_X,
+                handler: 'index.handler',
+                timeout: Duration.seconds(15),
+                encryption: this.props.encryption,
+                ...this.props.lambdaConfiguration
+            });
+
+            if (!Ec2AutomatedShutdown.policyByStack.has(stackId)) {
+                const policy = new PolicyStatement({
+                    actions: ['ec2:StopInstances'],
+                    effect: Effect.ALLOW
+                });
+                Ec2AutomatedShutdown.policyByStack.set(stackId, policy);
+                lambda.role.addToPolicy(policy);
+            }
+            Ec2AutomatedShutdown.lambdaByStack.set(stackId, lambda);
+        }
+        return Ec2AutomatedShutdown.lambdaByStack.get(stackId)!;
+    }
+
+    private addInstanceToPolicy(
+        instance: Instance | CfnInstance,
+        stack: Stack
+    ): void {
+        const stackId = stack.stackId;
+        const policy = Ec2AutomatedShutdown.policyByStack.get(stackId);
+
+        if (!policy) {
+            throw new Error('Policy not initialized');
+        }
+
+        const instanceArn = `arn:${Stack.of(instance).partition}:ec2:${stack.region}:${stack.account}:instance/${
```

Comment:
Prefer the use of `Aws.ACCOUNT`, `Aws.PARTITION`, and `Aws.REGION` which will synthesize down to CFN helpers and easier to read in the code